crypto/fipsmodule/rsa/rsa.c

```diff
@@ -1179,6 +1059,133 @@ int RSA_check_key(const RSA *key) {
+  // necessary for both FIPS checks performed here.
+  if (key_type == RSA_KEY_TYPE_FOR_CHECKING_INVALID ||
+      key_type == RSA_KEY_TYPE_FOR_CHECKING_PRIVATE_STRIP) {
+    OPENSSL_PUT_ERROR(RSA, RSA_R_BAD_RSA_PARAMETERS);
+    return 0;
+  }
+
+  // Validate basic properties of the key.
+  if (!RSA_check_key(key)) {
+    return 0;
+  }
+
+  BN_CTX *ctx = BN_CTX_new();
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  BIGNUM small_gcd;
+  BN_init(&small_gcd);
+
+  int ret = 0;
+  uint8_t *sig = NULL; // used later in the pair-wise consistency test.
+
+  // Perform partial public key validation of RSA keys (SP 800-89 5.3.3).
+  // Although this is not for primality testing, SP 800-89 cites an RSA
+  // primality testing algorithm, so we use |BN_prime_checks_for_generation| to
+  // match. This is only a plausibility test and we expect the value to be
+  // composite, so too few iterations will cause us to reject the key, not use
+  // an implausible one.
+  enum bn_primality_result_t primality_result;
+  if (BN_num_bits(key->e) <= 16 ||
+      BN_num_bits(key->e) > 256 ||
+      !BN_is_odd(key->n) ||
+      !BN_is_odd(key->e) ||
+      !BN_gcd(&small_gcd, key->n, g_small_factors(), ctx) ||
+      !BN_is_one(&small_gcd) ||
+      !BN_enhanced_miller_rabin_primality_test(&primality_result, key->n,
+                                               BN_prime_checks_for_generation,
+                                               ctx, NULL) ||
+      primality_result != bn_non_prime_power_composite) {
+    OPENSSL_PUT_ERROR(RSA, RSA_R_PUBLIC_KEY_VALIDATION_FAILED);
+    goto end;
+  }
+
+  // For public keys the check is over because the public key validation is
+  // the only thing we can test, we can't perform the pair-wise  consistency
+  // test without the private key.
+  if (key_type == RSA_KEY_TYPE_FOR_CHECKING_PUBLIC) {
+    ret = 1;
+    goto end;
+  }
+
+  // Only private keys (that contain the public component as well) can be
+  // tested with a pair-wise consistency test.
+  if (key_type != RSA_KEY_TYPE_FOR_CHECKING_PRIVATE_MIN &&
+      key_type != RSA_KEY_TYPE_FOR_CHECKING_PRIVATE &&
+      key_type != RSA_KEY_TYPE_FOR_CHECKING_PRIVATE_CRT) {
+      goto end;
+  }
+
+  // FIPS pair-wise consistency test (FIPS 140-2 4.9.2). Per FIPS 140-2 IG,
+  // section 9.9, it is not known whether |rsa| will be used for signing or
+  // encryption, so either pair-wise consistency self-test is acceptable. We
+  // perform a signing test. The same guidance can be found in FIPS 140-3 IG
+  // in Section 7.10.3.3, sub-section Additional comments.
+  uint8_t data[32] = {0};
+  unsigned sig_len = RSA_size(key);
+  sig = OPENSSL_malloc(sig_len);
+  if (sig == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_INTERNAL_ERROR);
+    goto end;
+  }
+
+  if (!RSA_sign(NID_sha256, data, sizeof(data), sig, &sig_len, key)) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_INTERNAL_ERROR);
+    goto end;
+  }
+  if (boringssl_fips_break_test("RSA_PWCT")) {
+    data[0] = ~data[0];
+  }
```

Comment:
NP: This will be optimized away for release build, but we could guard this logic with `#ifdef BORINGSSL_FIPS_BREAK_TESTS`,